### PR TITLE
drivers: entropy: Fix MCUX CAAM

### LINKED ISSUE #49440: test-ci: mimxrt11xx: testing.ztest.base.verbose_x and crypto.tinycryp : run failure no console output
**Describe the bug**
mimxrt11xx fsl_caam.c related failure including:

- testing.ztest.base.verbose_0
- testing.ztest.base.verbose_0_userspace
- testing.ztest.base.verbose_1
- testing.ztest.base.verbose_2
- tests/crypto/tinycryp
- tests/posix/eventfd
- drivers.entropy.entropy_get_entropy
- libraries.ring_buffer.capacity

Please also mention any information which could help others to understand
the problem you're facing:
west debug found caam init failure

```
#0  caam_out_job_ring_test_and_remove (found=<synthetic pointer>, wait=<synthetic pointer>, descriptor=<optimized out>, jobRing=kCAAM_JobRing0, base=<optimized out>)
    at /home/ubuntu/zephyr_project/zephyr_rt1160/modules/hal/nxp/mcux/mcux-sdk/drivers/caam/fsl_caam.c:568
#1  CAAM_Wait (base=base@entry=0x40440000, handle=handle@entry=0x800019a4 <z_interrupt_stacks+1828>, descriptor=descriptor@entry=0x80001880 <z_interrupt_stacks+1536>, mode=<optimized out>)
    at /home/ubuntu/zephyr_project/zephyr_rt1160/modules/hal/nxp/mcux/mcux-sdk/drivers/caam/fsl_caam.c:1860
#2  0x300066be in CAAM_RNG_Init (base=base@entry=0x40440000, handle=handle@entry=0x800019a4 <z_interrupt_stacks+1828>, stateHandle=<optimized out>, config=config@entry=0x8000199c <z_interrupt_stacks+1820>)
    at /home/ubuntu/zephyr_project/zephyr_rt1160/modules/hal/nxp/mcux/mcux-sdk/drivers/caam/fsl_caam.c:3463
#3  0x30006874 in CAAM_Init (base=0x40440000, config=config@entry=0x800019c8 <z_interrupt_stacks+1864>) at /home/ubuntu/zephyr_project/zephyr_rt1160/modules/hal/nxp/mcux/mcux-sdk/drivers/caam/fsl_caam.c:1773
#4  0x300050b8 in entropy_mcux_caam_init (dev=<optimized out>) at /home/ubuntu/zephyr_project/zephyr_rt1160/zephyr/drivers/entropy/entropy_mcux_caam.c:57
#5  0x30006ec4 in z_sys_init_run_level (level=level@entry=0) at /home/ubuntu/zephyr_project/zephyr_rt1160/zephyr/kernel/device.c:70
#6  0x300070fa in z_cstart () at /home/ubuntu/zephyr_project/zephyr_rt1160/zephyr/kernel/init.c:447
#7  0x30003c54 in z_arm_prep_c () at /home/ubuntu/zephyr_project/zephyr_rt1160/zephyr/arch/arm/core/aarch32/prep_c.c:260
#8  0x300042be in z_arm_reset () at /home/ubuntu/zephyr_project/zephyr_rt1160/zephyr/arch/arm/core/aarch32/cortex_m/reset.S:169

```

**To Reproduce**
cd tests/ztest/base
1. west build -b mimxrt1160_evk_cm7 -- -DCONF_FILE=prj_verbose_0.conf
2. west debug --runner jlink --tool-opt='-SelectEmuBySN 724879521' --  '--device=MIMXRT1166xxx6_M7' '--reset-after-load'

**Expected behavior**
test cases can run pass

**Logs and console output**

no console output

**Environment (please complete the following information):**
 - OS: (e.g. Linux, , )
 - Toolchain (e.g Zephyr SDK, ...)
 - Commit SHA or Version used:  zephyr-v3.1.0-4029-g8403866e39

### LINKED ISSUE #49846: mimxrt1160_evk network samples stopped working
Some commit after v3.1.0 broke the network sample apps when running on the mimxrt1160_evk board.  The problem is seen in the latest main branch, commit d5523386375af720facae445f50e6b1a1c7f7471.  It has also been reported this affects the mimxrt1170_evk as well.  Running the same sample from v3.1.0 boots and starts execution as expected.

**To Reproduce**
Steps to reproduce the behavior, using main branch:
```
west build -b mimxrt1160_evk_cm7 samples\net\cloud\tagoio_http_post --pristine
west flash
```
using the main branch, nothing prints to the console.  When building hello_world from the main branch, it prints to the console correctly.  

**Expected behavior**
using v3.1.0 release, after the same steps above, the below messages print to the console:
```
*** Booting Zephyr OS build zephyr-v3.1.0  ***


[00:00:00.505,737] <inf> net_config: Initializing network
[00:00:00.505,767] <inf> net_config: Waiting interface 1 (0x80002408) to be up...
```

### LINKED ISSUE #50446: MCUX CAAM is disabled temporarily
**Describe the bug**
The CAAM driver is currently disabled in zephyr because a change to the NXP HAL has caused it to fail its initialization. The problem is related to invalidating cache safely using cache write through mode, and fixing the issue requires setting up MPU regions with the correct settings.

**To Reproduce**
Currently not reproducible because the CAAM has been disabled by PR 49849 as damage control. On previous commits, it was reproducible by running any sample or application which used the entropy driver on rt11xx platforms.

**Expected behavior**
self-explanatory

**Impact**
The CAAM was a recently enabled feature (days before the HAL update in question), so this should not have much (if any) impact other than delaying the enablement of a feature which was recently upstreamed into the main zephyr branch.

**Logs and console output**
N/A

**Environment (please complete the following information):**
N/A

**Additional context**
none

### PR DESCRIPTION
**Pull request by decsny**
Fix the MCUX CAAM bug affecting RT11XX platforms when using the entropy subsystem

Fixes #50446
Fixes #49440
Fixes #49846

### CODE REVIEW

File: arch/arm/core/aarch32/Kconfig.mcux [hunk 1]
```diff
@@ -0,0 +1,5 @@
+# Copyright 2022 NXP
+# SPDX-License-Identifier: Apache-2.0
+
+config CPU_CORTEX_M
```
Comment (mmahadevan108) [inline arch/arm/core/aarch32/Kconfig.mcux:4]: Why is this needed since you have selected `DCACHE` in the `Kconfig.soc` file

Comment (decsny) [reply arch/arm/core/aarch32/Kconfig.mcux]: gone

**Comment by decsny** [comment]
@carlocaione I would like your opinion on the direction of some of the additions in this PR because it relates to an area which you were active in.

In order to enable this driver for the CAAM, I had to make use of nocache memory. The RT11XX has a cortex M4 target that has a kinetis cache. The ARM initialization code in zephyr seems to assume that if there is a cache, the target is an M7, because it calls SCB APIs to clean and invalidate the DCACHE. Because these functions are undefined for the M4 target, I had to specify that this cache is not an ARM cache. I used the HAS_EXTERNAL_CACHE kconfig to do this because it seemed appropriate, but since that option depends on the CACHE_MANAGEMENT symbol, I had to write an empty implementation for a Kinetis cache even though it isnt't really used, just in order to satisfy the kernel common test that twister reported would fail if those APIs were not defined. What are your thoughts on this?

**Comment by carlocaione** [comment]
> The ARM initialization code in zephyr seems to assume that if there is a cache, the target is an M7, because it calls SCB APIs to clean and invalidate the DCACHE.

This is the base problem. The ARM initialization code must be fixed to not use SCB API calls for cache if `CONFIG_DCACHE` is not defined. See for example: https://github.com/zephyrproject-rtos/zephyr/blob/1d879b93c59e8f1f48d86db5794600769c86facb/arch/arm/core/aarch32/cortex_m/scb.c#L111-L122

So in your case your SOC must not define `CONFIG_CPU_HAS_DCACHE / CONFIG_DCACHE` and all the call to the SCB should be skipped.

**Comment by decsny** [comment]
> 

Yes but I need to have ARCH_HAS_NOCACHE_MEMORY enabled in order to use the NOCACHE_MEMORY Kconfig, since this driver requires use of nocache memory, even though it is used on an M4, because NXP has a custom architecture that doesn't quite exactly follow ARM spec. ARCH_HAS_NOCACHE_MEMORY depends on the DCACHE kconfig.

**Comment by decsny** [comment]
I am going to work on a separate PR for this M4 cache problem, I am declaring this PR inactive for now from my side

**Comment by decsny** [comment]
This PR now will depend on #50871

**Comment by carlocaione** [comment]
> I found a better way to do this, so this PR now will depend on #50871

I commented there.

**Comment by decsny** [comment]
update: the caching problem on m4 is too much of a bottleneck, disabled in devicetree, for now m7 can at least be supported

File: tests/drivers/entropy/api/src/main.c [hunk 2]
```diff
@@ -25,6 +25,11 @@
 #define BUFFER_LENGTH           10
 #define RECHECK_RANDOM_ENTROPY  0x10
 
+#ifdef CONFIG_RANDOM_BUFFER_NOCACHED
+__attribute__((__section__(".nocache")))
+static uint8_t buffer[BUFFER_LENGTH] = {0};
```
Comment (danieldegrasse) [inline tests/drivers/entropy/api/src/main.c:30]: Just a suggestion: move the definition of `buffer` here for targets where `CONFIG_RANDOM_BUFFER_NOCACHED` won't be selected, to improve readability. Something like this:

```
#ifdef CONFIG_RANDOM_BUFFER_NOCACHED
__attribute__((__section__(".nocache")))
static uint8_t buffer[BUFFER_LENGTH] = {0};
#else
static uint8_t buffer[BUFFER_LENGTH] = {0};
#endif
```

Comment (decsny) [reply tests/drivers/entropy/api/src/main.c:30]: done

File: tests/drivers/entropy/api/Kconfig.mcux [hunk 3]
```diff
@@ -0,0 +1,9 @@
+# Copyright NXP 2022
```
Comment (mmahadevan108) [inline tests/drivers/entropy/api/Kconfig.mcux:1]: Minor point; missing a comma after NXP

File: tests/drivers/entropy/api/Kconfig [hunk 4]
```diff
@@ -0,0 +1,10 @@
+# Copyright NXP 2022
```
Comment (mmahadevan108) [inline tests/drivers/entropy/api/Kconfig:1]: Minor point, Missing a comma after NXP

File: tests/drivers/entropy/api/Kconfig.mcux (hunk 3 above)
Comment (mmahadevan108) [inline tests/drivers/entropy/api/Kconfig.mcux:1]: This should be selected in the boards folder config file.

File: tests/drivers/entropy/api/Kconfig [hunk 5]
```diff
@@ -0,0 +1,10 @@
...
+# SPDX-License-Identifier: Apache-2.0
+
+mainmenu "Entropy API Test"
+
+source "Kconfig.zephyr"
+source "tests/drivers/entropy/api/Kconfig.mcux"
```
Comment (mmahadevan108) [inline tests/drivers/entropy/api/Kconfig:7]: This can be deleted

Comment (decsny) [reply tests/drivers/entropy/api/Kconfig:7]: I am confused what you mean, what can be deleted and why?

File: tests/drivers/entropy/api/Kconfig.mcux (hunk 3 above)
Comment (decsny) [reply tests/drivers/entropy/api/Kconfig.mcux:1]: I am not sure if I agree, because this is needed because of using the caam, it's not specific to a board

Comment (mmahadevan108) [reply tests/drivers/entropy/api/Kconfig.mcux:1]: I am wondering about future SoC variations from NXP that may not need this Kconfig.

Comment (decsny) [reply tests/drivers/entropy/api/Kconfig.mcux:1]: okay.... but what about future variations that would need this kconfig because they have a CAAM... you would need to remember to add a board overlay on this test for every evk with a caam.... it just seems easier to have the kconfig selected automatically just depending of if the SOC has a caam

**Review by mmahadevan108 — APPROVED** [review]

**Review by danieldegrasse — APPROVED** [review]

**Review by ceolin — APPROVED** [review]
